File: client/python/mesh/system/mesh_builtin.py
#
# Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
# TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
#
#
import json
import os
import platform
from typing import List, Dict

import mesh.tool as tool
from mesh.kinds import Versions
from mesh.macro import spi, mps, ServiceLoader
from mesh.prsim import Builtin, Hodor


@mps
@spi("mesh")
class MeshBuiltin(Builtin):

    def doc(self, name: str, formatter: str) -> str:
        return ""

    def version(self) -> Versions:
        def prop() -> Dict[str, str]:
            try:
                with open(os.path.join(tool.pwd(), f"{tool.get_mesh_name()}.version")) as vfd:
                    vps = json.load(vfd)
                    return vps if vps else {}
            except FileNotFoundError:
                return {}

        properties = prop()
        name = f"{tool.get_mesh_name()}_version".upper()
        mv = tool.get_property("", [name])
        version = tool.anyone(mv, properties.get("version", "1.0.0"))
        versions = Versions()
        versions.version = version
        versions.infos = {
            f"{tool.get_mesh_name()}.commit_id": properties.get("commit_id", "3dd81bc"),
            f"{tool.get_mesh_name()}.os": platform.system(),
            f"{tool.get_mesh_name()}.arch": platform.architecture()[0],
            f"{tool.get_mesh_name()}.version": version,
        }
        return versions

    def debug(self, features: Dict[str, str]):
        doors = ServiceLoader.load(Hodor).list('')
        if not doors:
            return
        for hodor in doors:
            hodor.debug(features)

    def stats(self, features: List[str]) -> Dict[str, str]:
        indies = {}
        doors = ServiceLoader.load(Hodor).list('')
        if not doors:
            return indies
        for hodor in doors:
            stats = hodor.stats(features)
            if not stats:
                continue
            for key, value in stats.items():
                indies[key] = value
        return indies

File: client/java/src/main/java/io/be/mesh/system/MeshBuiltin.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package io.be.mesh.system;

import io.be.mesh.macro.Index;
import io.be.mesh.macro.MPS;
import io.be.mesh.macro.SPI;
import io.be.mesh.mpc.Codec;
import io.be.mesh.mpc.ServiceLoader;
import io.be.mesh.mpc.Types;
import io.be.mesh.prsim.Builtin;
import io.be.mesh.prsim.Hodor;
import io.be.mesh.struct.Versions;
import io.be.mesh.tool.Tool;

import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * @author coyzeng@gmail.com
 */
@MPS
@SPI("mesh")
public class MeshBuiltin implements Builtin {

    @Override
    public String doc(String name, String formatter) {
        ByteBuffer buffer = ServiceLoader.resource(String.format("%s.schema", Tool.MESH_NAME.get()));
        Codec json = ServiceLoader.load(Codec.class).get(Codec.JSON);
        Codec codec = ServiceLoader.load(Codec.class).get(formatter);
        return codec.encodeString(json.decode(buffer, Types.MapObject));
    }

    @Override
    public Versions version() {
        ByteBuffer buffer = ServiceLoader.resource(String.format("%s.version", Tool.MESH_NAME.get()));
        Codec json = ServiceLoader.load(Codec.class).get(Codec.JSON);
        Versions versions = json.decode(buffer, Types.of(Versions.class));
        String name = Tool.toUpperCase(String.format("%s_version", Tool.MESH_NAME.get()));
        String mv = Tool.getProperty("", name);
        if (null != versions && Tool.required(mv)) {
            versions.setVersion(mv);
            Optional.ofNullable(versions.getInfos()).ifPresent(x -> x.put(name, mv));
        }
        return versions;
    }

    @Override
    public void debug(List<String> features) {
        List<Hodor> doors = ServiceLoader.load(Hodor.class).list();
        if (Tool.optional(doors)) {
            return;
        }
        for (Hodor hodor : doors) {
            hodor.debug(features);
        }
    }

    @Override
    public Map<String, String> stats(@Index(0) List<String> features) {
        List<Hodor> doors = ServiceLoader.load(Hodor.class).list();
        if (Tool.optional(doors)) {
            return new HashMap<>(0);
        }
        Map<String, String> indies = new HashMap<>();
        for (Hodor hodor : doors) {
            Map<String, String> stats = hodor.stats(features);
            if (Tool.required(stats)) {
                indies.putAll(stats);
            }
        }
        return indies;
    }

}
